### ISSUE TITLE
[UX] When a new content type is created, automatically assign permissions to the Editor role.

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#5885]
## Description of the need

This is a sibling issue to #5350, and this is a related issue against Drupal 7 core: [Dynamic permissions cannot be automatically assigned to or removed from roles.](https://www.drupal.org/project/drupal/issues/787152)

The 3 content types that come OOTB with a default Backdrop installation all have these permissions already granted to the Editor role, but that is not the case when creating new content types:
![image](https://user-images.githubusercontent.com/2423362/206680040-694dde0c-9aa9-4427-bc16-d19289132d05.png)

When creating a new content type, these permissions need to be manually adjusted each time. With #900, we have improved the UX in Backdrop, since you don't need to navigate away from the content type creation/edit form in order to assign the required permissions to the various roles, nor do you have to search through a HUGE list of permissions to find the correct ones 👍🏼 ...this issue here is to improve the UX a big further, by allowing some sensible defaults (that way, site builders will have to scroll/click less when configuring content types).

## Proposed solution

- Automatically grant these permissions to the Editor role.
- Perhaps add a setting for this in `admin/config/people/roles`, similar to the "The role that will be automatically assigned new permissions whenever a module is enabled:" setting for admin roles.
- Also throw a notice/info message in the node creation form, letting the user know that automatic permissions assignment has happened, and that they might want to review/tweak them as needed.

---
Advocate for this feature: @klonos 
---

Related issues
* https://github.com/backdrop/backdrop-issues/issues/5031 
* https://github.com/backdrop/backdrop-issues/issues/1285
* https://github.com/backdrop/backdrop-issues/issues/6078

**Comment by klonos** [comment]
PR available here: https://github.com/backdrop/backdrop/pull/4273

`admin/config/people/roles` before:
![image](https://user-images.githubusercontent.com/2423362/206785311-cecd17bc-04a0-49d1-b82f-9d8ce0ac4f30.png)

`admin/config/people/roles` after (only in new sites - the default editor in existing sites is set to "disabled" in order to retain current functionality):
![image](https://user-images.githubusercontent.com/2423362/206785491-3ccae3fc-ced1-486e-a6da-f5855ea1bb0b.png)

Node creation form (only on new content types):
![image](https://user-images.githubusercontent.com/2423362/206785001-376941ee-adae-40d5-beb6-9d399074f145.png)

**Comment by klonos** [comment]
I went to get some sleep, and then these thoughts came to mind:

1. What if the person creating the content type is some kind of "junior admin" or the site owner? In these cases, the site builder/administrator might have created a "lesser" user role for them, that allows them to do certain things, but not to be configuring permissions. What would/should happen then?
2. What if a security-cautious organization prefers to disable the auto-assignment of permissions (even for the admin role), and prefers to only be assigning permissions after very careful consideration and site audits instead?

If a person in this scenario creates a content type, they will end up in a situation when nobody can view or create content of this type, end they will have no way to know that beforehand.

So here's the use case I've tested on my local:
- I've set the auto-permission assignment in `admin/config/people/roles` to "disabled" for both the default admin, as well as the default content editor.
- I've created a "Junior admin" role, with the following permissions (no permission to manage roles/permissions):
  - access administration bar
  - administer fields
  - administer content types
  - administer nodes
  - access administration pages
  - view the administration theme
- I've added a new user account, and only gave them that user role.
- I've logged into the local site as that user, and tried to create a new content type

The result of the scenario above was that:
- no "Permissions" vertical tab was available for that user while creating a new content type (as expected 👍🏼 )
- no permissions were set for that content type (expected 👍🏼 )
- of course, nobody could create/view/edit content of that type - only user1 (expected 👍🏼 )
- no warning or indication of that was shown to the user, so they are left wondering what's going on after they have created the new content type (no good UX 👎🏼 )
- user1 could still go to `admin/config/people/permissions` and assign permissions accordingly, so no "hard-locking" here.

So I've tweaked the PR to show helpful messages in the following scenarios, **if the user that's creating the content type has no access to configure permissions**:
- only the default admin role has been configured for automatic permission assignment (existing sites):
  ![image](https://user-images.githubusercontent.com/2423362/206828282-4a52d103-7c05-449d-9a74-da2ce8514dd2.png)
- both the default admin role and the default content editor roles have been configured for automatic permission assignment (new sites when this feature gets in):
  ![image](https://user-images.githubusercontent.com/2423362/206828295-a9d167a8-4a4f-4036-95b1-665d1ef910d7.png)
- both of the default admin/editor roles have been intentionally disabled:
  ![image](https://user-images.githubusercontent.com/2423362/206828325-4410034f-eaa9-4ae4-b8ba-78a472f915ef.png)

And here's what user1 will see on sites when both of the default admin/editor roles have been intentionally disabled:
![image](https://user-images.githubusercontent.com/2423362/206835330-362fe2db-01b2-47bd-acef-7458507a3c8b.png)

PS: All _info_ messages in this PR are shown only when creating a new content type, while _warnings_ of no permissions being set are shown for existing content types as well (to help people when troubleshooting).

**Comment by kiamlaluno** [comment]
The warnings assume that disabling the admin and editor roles means nobody can create, view, or edit nodes of that content type, but that isn't true: Disabling the admin and editor roles just means there aren't roles to which permissions are automatically granted.

The info messages are correct, but are they giving useful information to the user who can create content types, but who cannot change permissions?
Knowing that nobody can create nodes of that content type is helpful information, but when the user can change the permissions; in that case, it's a reminder the user needs to set new permissions.

**Comment by kiamlaluno** [comment]
What I meant is that users with the *bypass node access* permission can create, view, or edit nodes for any content type; other users could get any of those permissions from a `hook_node_access()` implementation. 

Saying that *nobody will be able to create, view, or edit any content of this type* when the *Administrator* or *Editor* role has been disabled isn't exact.

Saying that *the Administrator and the Editor roles have been automatically assigned some sensible default permissions to be able to create, view, or edit any content of this type* is correct, but it's probably irrelevant, given there are other users who can create, view, or edit nodes of the newly created content type.

**Comment by kiamlaluno** [comment]
[`node_node_access()`](https://docs.backdropcms.org/api/backdrop/core%21modules%21node%21node.module/function/node_node_access/1) is the implementation that checks for the `"create $type content"` permission. It's sufficient another `hook_node_access()` implementation returns [`NODE_ACCESS_DENY`](https://docs.backdropcms.org/api/backdrop/core%21modules%21node%21node.module/constant/NODE_ACCESS_DENY/1) to make that permission "inefficient." In fact, [`Node::createAccess()`](https://docs.backdropcms.org/api/backdrop/core%21modules%21node%21node.entity.inc/function/Node%3A%3AcreateAccess/1) contains the following code.

```php
if (in_array(NODE_ACCESS_DENY, $access, TRUE)) {
    $rights[$account->uid][$bundle] = FALSE;
    return $rights[$account->uid][$bundle];
  }
  elseif (in_array(NODE_ACCESS_ALLOW, $access, TRUE)) {
    $rights[$account->uid][$bundle] = TRUE;
    return $rights[$account->uid][$bundle];
  }
```

Checking the permissions isn't generally sufficient to say whether no users can create nodes of the newly created content type.

**Comment by kiamlaluno** [comment]
On the preview site, I created a new content type, disabled the *Default administrator role* and the *Default content editor role*, and edited the new content type. I got the *Nobody will be able to create, view or edit any content of this type until permissions have been configured appropriately for it.* message, despite the *Administrator* user still have the permission to edit, create, and delete nodes of the new content type. Furthermore, since I logged in as user #1, I also have the *bypass node access* permission, and I am able to edit, create, and delete nodes of the new content type for that reason too.

**Comment by klonos** [comment]
Hmm 🤔 ...gotcha @kiamlaluno 👍🏼 (and thanks for taking the time to elaborate)

I'm thinking to strip the whole messages thing out of the current PR, and only leave the code related to the auto-assignment of permissions. We could file a follow-up for the messages (if it actually makes sense to have them, given the above discussion).

**Comment by kiamlaluno** [comment]
@klonos I would leave the part about messages in a follow-up. Having the permissions automatically given to the *Editor* role is, IMO, a useful feature, either in the sites where the users who create a content type aren't allowed to change the permissions assigned to roles, or in sites where the users who create new content type are also allowed to change role permissions. (It's more probable that users who can create new content types aren't allowed to change permissions.)

**Comment by kiamlaluno** [comment]
Checking if there are roles with the permission to create nodes of a specific content type should just require to check the value returned from `user_roles(FALSE, "create $type content")` which returns an array of role labels with that permission; if the array is empty, no roles have that permission. Eventually, I would also check the value returned from `user_roles(FALSE, 'bypass node access')` to include those users who can create nodes of a specific content type without having an explicit permission.
Since checking that doesn't give the certainty there aren't users who can create nodes of the new content type, I would change the warning message to say that probably there aren't users who cannot create nodes of that content type.

At this point, I am not sure if splitting the issue is really necessary; leaving the message part to another issue would probably get the *automatically assign permissions to the "Editor" role* part quickly in.

**Comment by docwilmot** [comment]
Good issue. I would suggest just a message immediately upon creation of a new content type, saying that the Editor role has been granted permissions to manage this content type, and leave it at that.

**Comment by klonos** [comment]
Thanks for the feedback @docwilmot 🙏🏼 ...the way I have currently implemented the feature is that the permissions for whatever role has been configured as the "default editor" are pre-selected when creating the content type. That doesn't prevent the person creating the content type from switching to the "Permissions" tab, and undoing this pre-selection before saving.

In your comment, you say "a message immediately upon creation of a new content type", which makes me think that you believe that the permissions should be automatically saved, regardless (which means that I'd also need to change the current implementation). Did I get that right?

**Comment by docwilmot** [comment]
Actually, I'd forgotten about the vertical fieldsets for permissions on type creation page. Settings there however we agree should be expected defaults and users shouldn't *need* to check; would we expect users to know this permission is set for editors? Maybe if the defaults are unchanged on save we set the message I mentioned above?

**Comment by klonos** [comment]
> Maybe if the defaults are unchanged on save we set the message I mentioned above?

Gotcha 👍🏼

**Comment by stpaultim** [comment]
I was just testing this issue. I like what it's doing, but I think we can make the UI and text better. I'll start here by trying to simply some of the text/labels.

WAS:
> Administrator and Editor roles have been automatically assigned some sensible default permissions, to be able to create, view or edit any content of this type. You may want to further tweak these in the Permissions tab below. This behavior can be adjusted in the Automatic permission assignment section of the [Roles] confirguration page.

SUGGESTION:
> Administrator and Editor roles have been assigned permission to create, view, or edit any content of this type, please be sure to make necessary adjustments. Which roles are automatically assigned permissions can be set on the [Roles} configuration page.

-----

![image](https://user-images.githubusercontent.com/3144571/231398277-31d73645-c052-4843-9176-697b071924e4.png)

WAS:
> Whenever a new content type is created, permissions required to create, edit and manage content of that type will be automatically granted to the role selected here.

SUGGESTION:
> This role is automatically given permission to create, edit, and manage content for any newly created content types.

**Comment by klonos** [comment]
Thanks for the suggestions @stpaultim 🙏🏼 ...they seem much more simplified/streamlined and certainly less wordy. I've adjusted the PR, although I've not used your suggestions 100% verbatim - I tweaked them, and used them as inspiration.

Here's what's changed:
- I've renamed the fieldset to "Default roles", as suggested 👍🏼 
- I've changed instances of `create, view, or edit` to `create, edit, or delete`, since that is the set of permissions ("view" is handled by published/unpublished etc.).
- I've shortened the main warning/info message shown above the vertical tabs, by moving the `You can change which roles are automatically assigned permissions...` message inside the "Permissions" tab (acting as a tip).
- The `, please be sure to make necessary adjustments.` suggestion wouldn't work as it was (part of the first sentence), as it needs to be conditionally shown only when people have permission to manage permissions.
- Since the "Permissions" tab is not the default tab in the form, it is necessary to point people to it, so I kept the instructions on where to locate the settings: `You can make any necessary adjustments in the _Permissions_ tab below.`
- I kept the `in the "Default roles" section` portion of the text, since that fieldset is collapsed by default, so not immediately obvious where these settings are when someone navigates to the role configuration page.

Some screenshots:
- What people will see by default when creating a new content type:
  ![image](https://user-images.githubusercontent.com/2423362/231784607-f6b8d0af-0414-4eaa-9014-403f8932635d.png)
- Tip shown when switching to the "Permissions" vertical tab:
  ![image](https://user-images.githubusercontent.com/2423362/231785177-9f38e85d-6af0-4de8-892e-fc9806fc7634.png)
- Configuration options (collapsed by default - expanded here):
  ![image](https://user-images.githubusercontent.com/2423362/231784096-be4454a6-c20d-4138-9e22-0e93acd9a0c5.png)
- What people will see if the content type has no permissions set (can happen on existing sites before this feature):
  ![image](https://user-images.githubusercontent.com/2423362/231785581-6582c1a3-dbae-4781-a632-0b9edd6fe9bc.png)
- Same scenario as above, but the current user doesn't have access to manage permissions:
  ![image](https://user-images.githubusercontent.com/2423362/231788654-b92db1b1-8553-49b6-8644-64b4bda692f6.png)

**Comment by klonos** [comment]
PS: I've included the changes from the PR in https://github.com/backdrop/backdrop/pull/4395 in order to have tests pass the cspell check - I will revert that change once that PR is merged.

**Comment by klonos** [comment]
> What do you think about displaying the second text (see below) as normal or 'description' text?

I thought of that, but usually description/help text is "tied" to a specific form element, providing information directly related to what the element does, or instructions on what the value/format of the field should be. That's why in this instance I had it styled as a "tip" instead, because it is not directly related to the permission checkboxes in that tab. Anyway...

> I would rather leave out the message about the Default roles section.

Done 👍🏼 

I've also:
- rebased the PR branch, to remove the cspell workaround since #6047 got merged (thanks @herbdool 🙏🏼 ).
- tweaked the wording of the warning when no permissions have been assigned for a content type ("nobody will be able" was not 100% accurate):
  ![image](https://user-images.githubusercontent.com/2423362/231913400-81e6a1aa-d64e-4519-a3ce-bf6ff4a5b1d2.png)
  ![image](https://user-images.githubusercontent.com/2423362/231913735-bf0d2074-5fc8-4458-948b-1d8b87f36e9c.png)

**Comment by stpaultim** [comment]
I would say this is "Works for Me" for me. But, I'd like to hear at least one other person agree before changing the status, given the amount text changes, etc.

**Comment by olafgrabienski** [comment]
Works for me, too!

**Comment by jenlampton** [comment]
I saw this today and it immediately made me nervous about how much access we are granting editors. On most of my sites where I have editors, I don't give them permission to delete anything -- especially not anything created by anyone else. I would like to propose that we remove the "Delete all X" permission from those that are automatically granted to editors.

**Comment by kiamlaluno** [comment]
If editors are supposed to only edit content (as the name implies), then effectively deleting content is too much. That would probably be a task for *content moderators* or *site moderators*.

**Comment by jenlampton** [comment]
I have opened https://github.com/backdrop/backdrop-issues/issues/6078 to address the defaults that are set for page and post.

**Comment by quicksketch** [comment]
With @klonos the advocate we can move this into the minor milestone.

**Comment by quicksketch** [comment]
I gave this a review and it looks good! There are a few issues that should be addressed:

- We should add an update hook to specify `user_editor_role`, even if it's disabled on existing sites.
- Test coverage. There's an existing test `testAdministratorRole` in `user.test` that might be used or need to be adapted as well.

I wasn't sure about the messages based on screenshots, but in testing the sandbox I think they're a useful addition. As a follow-up, I'd like if we could make some official solution for inline messages to differentiate them (if needed) from the top-of-page messages.

**Comment by stpaultim** [comment]
I created a PR for: https://github.com/backdrop/backdrop-issues/issues/6078

**Comment by quicksketch** [comment]
@klonos I pushed up test coverage to your PR. Other than adding the tests, the only thing I changed was adding an Oxford comma to "create, edit and delete" (now "create, edit, and delete").

I manually tested the functionality as part of writing the tests and did not encounter any issues.

**Comment by klonos** [comment]
As per [discussion on Zulip](https://backdrop.zulipchat.com/#narrow/stream/218635-Backdrop/topic/Feature.20Freeze/near/355039602), given that both @olafgrabienski and @stpaultim gave this a "WFM" in earlier comments here, and since the only change since then was the addition of the update hook and the tests, marking this accordingly.

**Comment by quicksketch** [comment]
With all tests passing and previous approvals on the functionality, I've merged https://github.com/backdrop/backdrop/pull/4273 into 1.x for 1.25.0. I think this is the last feature we'll see in this release. Thanks everyone for their work!

https://github.com/backdrop/backdrop/commit/4074679e83b3609d57388713784c370f873b2ff9 by @klonos, @quicksketch, @kiamlaluno, @docwilmot, @stpaultim, and @olafgrabienski.

**Comment by jenlampton** [comment]
It looks like we forgot to remove the "delete any" permission for all node types. We should try to get that change in before the 1.25.0 release.

**Comment by jenlampton** [comment]
Update: it was not very hard. There is a PR ready for review/testing if anyone is interested:  https://github.com/backdrop/backdrop/pull/4428
